// type-chef-di/chainingOptions/InstantiationModeCO.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstantiationModeCO = void 0;
const ChainingOptions_1 = require("./ChainingOptions");
const AsPrototypeCO_1 = require("./AsPrototypeCO");
const AsFactoryResultCO_1 = require("./AsFactoryResultCO");
const AsFactoryCO_1 = require("./AsFactoryCO");
class InstantiationModeCO {
    constructor(container, key) {
        this.container = container;
        this.key = key;
        this.instantiable = this.container.definitionsRepository.getDefinition(key);
        this.chainingOptions = new ChainingOptions_1.ChainingOptions(container, key);
    }
    asPrototype() {
        this.chainingOptions.asPrototype();
        return new AsPrototypeCO_1.AsPrototypeCO(this.container, this.key);
    }
    asSingleton() {
        this.chainingOptions.asSingleton();
        return new AsPrototypeCO_1.AsPrototypeCO(this.container, this.key);
    }
    asConstant() {
        this.chainingOptions.asConstant();
    }
    addTags(tags) {
        this.chainingOptions.addTags(tags);
    }
    asFactoryResult(factoryKey) {
        this.chainingOptions.asFactoryResult(factoryKey);
        return new AsFactoryResultCO_1.AsFactoryResultCO(this.container, this.key);
    }
    asFactory(factoryFn) {
        this.chainingOptions.asFactory(factoryFn);
        return new AsFactoryCO_1.AsFactoryCO(this.container, this.key);
    }
    withContext(context) {
        this.chainingOptions.withContext(context);
    }
}
exports.InstantiationModeCO = InstantiationModeCO;
