// type-chef-di/definitions/FactoryInstantiation.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FactoryInstantiation = void 0;
const ConstructorInstantiation_1 = require("./ConstructorInstantiation");
class FactoryInstantiation {
    constructor(definition, resolver) {
        this.tags = [];
        this.definition = definition;
        this.constructorInstantiation = new ConstructorInstantiation_1.ConstructorInstantiation({
            key: definition.key,
            content: definition.content,
            context: definition.context,
            instantiationMode: definition.instantiationMode
        }, resolver);
    }
    instantiate() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.constructorInstantiation.instantiate();
        });
    }
}
exports.FactoryInstantiation = FactoryInstantiation;
